### PR TITLE
Box slits and echelle order matching

### PR DESCRIPTION
**Pull request by kbwestfall**
This builds on and supersedes #964 .

The two main things this PR does are:

- Introduce a way of identifying box slits, using an added parameter to `EdgeTracePar`.  This was meant as a way of distinguishing between erroneous slits and slits that are useful for calculating the flexure.

- Forces all echelle spectrographs to match the order to the left-right edge pairs when constructing the `SlitTraceSet` object (using the pre-existing `ech_order` part of the datamodel).  Because this matching is set at the outset and doesn't change, this removes the need for the `order_vec` and `slit2order` methods in `spectrograph`, as well as essentially making the `match_tol_spat_pos` property from #964 an element of `EdgeTracePar`.

Changes are significant enough that I'll be running the dev-suite and reconstructing cooked.  Some testing is still ongoing, but the code is ready for review.

### CODE REVIEW

File: pypeit/edgetrace.py [hunk 1]
```diff
@@ -169,9 +169,14 @@ def __init__(self):
                    ('MASKINSERT', 'Trace was inserted based on drilled slit-mask locations'),
                  ('ORPHANINSERT', 'Trace was inserted to match an orphaned edge'),
                     ('SHORTSLIT', 'Slit formed by left and right edge is too short'),
```
Comment (profxj) [inline pypeit/edgetrace.py:171]: maybe add 'to be valid'.

File: pypeit/edgetrace.py [hunk 2]
```diff
@@ -2479,40 +2650,38 @@ def remove_traces(self, indx, resort=True, rebuild_pca=False, sync_rm='ignore'):
...
-                msgs.info('Removing both traces if either selected in synchonized pair.')
-            elif sync_rm == 'neither':
-                _indx = np.repeat(np.all(_indx.reshape(-1,2), axis=1), 2)
-                msgs.info('Removing both traces only if both selected in synchonized pair.')
-            else:
-                msgs.error('Unknown sync removal keyword: {0}'.format(sync_rm))
+        # Check the input
+        _indx = self.synced_selection(np.atleast_1d(indx), mode=sync_rm)
+
+#        # Deal with removing traces when they are left-right
```
Comment (profxj) [inline pypeit/edgetrace.py:2673]: remove?

File: pypeit/spectrographs/gemini_gnirs.py [hunk 3]
```diff
@@ -328,6 +333,8 @@ def orders(self):
 
     @property
     def spec_min_max(self):
+        # TODO: Why aren't these numbers in fraction of the detector
```
Comment (profxj) [inline pypeit/spectrographs/gemini_gnirs.py:336]: Good question, although I don't think you can bin near-IR detectors.

File: pypeit/edgetrace.py (hunk 1 above)
Comment (kbwestfall) [reply pypeit/edgetrace.py]: Done.

File: pypeit/edgetrace.py (hunk 2 above)
Comment (kbwestfall) [reply pypeit/edgetrace.py]: Done.

File: pypeit/spectrographs/gemini_gnirs.py (hunk 3 above)
Comment (kbwestfall) [reply pypeit/spectrographs/gemini_gnirs.py:336]: Right.  It's mostly just a comment about consistency, given that most (all?) of the other positions are given as relative to the detector size (for all echelles, regardless of their wavelength range).  Happy to leave it or change it and propagate the change for all spectrographs.

**Comment by kbwestfall** [comment]
> Looks great. I'll just ask that we ponder whether the
> order assignment is sufficiently general that when
> we go to non-fixed format echelle spectrographs (e.g. HIRES)
> they will be ok to implement.

Yeah, I need input from you all on this.  What I like about the new approach is that you only have to associate "slit" traces from the `trace` frames to echelle orders once, and that is all contained in `EdgeTraceSet.match_order`.  The current code allows you to adjust the match based on the specific observations using an offset (which it doesn't try to optimize), and then gives you a report that looks like this (for the example `gemini_gnirs` data @jhennawi  sent me):

```
[INFO]    :: Before vetting, the echelle order, matching left-right trace pair, and matching separation are:
[INFO]    ::   ORDER PAIR    SEP
[INFO]    ::  ------ ---- ------
[INFO]    ::       6    1  0.022
[INFO]    ::       5    2 -0.021
[INFO]    ::       4    3 -0.019
[INFO]    ::       3    4 -0.023
[INFO]    ::  ------ ---- ------
```

The details of the order matching can be changed within `match_order` and none of the rest of the code has to change.  The core assumption is that, regardless of how the exact edge locations are changed (illumination profile tweaks and flexure changes), the gross association of order ID and "slit"/trace ID never changes.  If that core assumption is wrong, then we'll need to rethink this.

**Comment by kbwestfall** [comment]
Working to fix these:
```
--- PYPEIT DEVELOPMENT SUITE FAILED 11/51 TESTS (Masters ignored) ---
Failed tests:
    run_pypeit: keck_lris_blue/multi_600_4000_d560
    run_pypeit (with masters): keck_lris_red/multi_1200_9000_d680_1x2
    run_pypeit (with masters): keck_lris_red/multi_600_5000_d560
    run_pypeit (with masters): keck_lris_red/multi_1200_9000_d680
    run_pypeit (with masters): keck_lris_red/multi_400_8500_d560
    run_pypeit: vlt_xshooter/VIS_1x1
    run_pypeit: vlt_xshooter/VIS_2x2
    run_pypeit: vlt_xshooter/NIR
    run_pypeit: gemini_gmos/GS_HAM_R400_860
    run_pypeit: magellan_mage/1x1
    run_pypeit: lbt_luci/LUCI-II
```

**Comment by kbwestfall** [comment]
Tests pass with new Cooked.

```
test_alignments.py .                                                     [  0%]
test_arc.py .                                                            [  0%]
test_arcimage.py ..                                                      [  1%]
test_biasframe.py ..                                                     [  2%]
test_bitmask.py ....                                                     [  4%]
test_bpmimage.py ..                                                      [  5%]
test_bspline.py ..........                                               [  9%]
test_buildimage.py .                                                     [  9%]
test_calibrations.py .......                                             [ 12%]
test_chk_calibs.py ..                                                    [ 13%]
test_coadd.py .                                                          [ 14%]
test_cooked.py .                                                         [ 14%]
test_coremeta.py .                                                       [ 15%]
test_datacontainer.py ........                                           [ 18%]
test_detector.py ...                                                     [ 19%]
test_edgetrace.py .                                                      [ 20%]
test_flatfield.py .                                                      [ 20%]
test_flexure.py .                                                        [ 21%]
test_flux.py ....                                                        [ 22%]
test_fluxspec.py ..                                                      [ 23%]
test_load.py .                                                           [ 24%]
test_load_images.py ...........                                          [ 29%]
test_masterframe.py .                                                    [ 29%]
test_match.py ....                                                       [ 31%]
test_metadata.py .....                                                   [ 33%]
test_moment.py ......                                                    [ 35%]
test_msgs.py ..                                                          [ 36%]
test_opticalmodel.py ....                                                [ 38%]
test_parse.py ...                                                        [ 39%]
test_pca.py ....                                                         [ 41%]
test_processimage.py ..                                                  [ 42%]
test_procimg.py .                                                        [ 42%]
test_pydl.py .                                                           [ 43%]
test_pypeit.py ..                                                        [ 44%]
test_pypeitimage.py .                                                    [ 44%]
test_pypeitpar.py .................................                      [ 58%]
test_pypeitsetup.py .......                                              [ 61%]
test_qa.py .                                                             [ 62%]
test_runpypeit.py ..                                                     [ 63%]
test_scienceimage.py ...                                                 [ 64%]
test_scripts.py .........                                                [ 68%]
test_setups.py ................                                          [ 75%]
test_skysub.py ..                                                        [ 76%]
test_slitmask.py .                                                       [ 76%]
test_slittrace.py ....                                                   [ 78%]
test_spec2dobj.py ......                                                 [ 80%]
test_specobj.py ......                                                   [ 83%]
test_specobjs.py .....                                                   [ 85%]
test_spectrographs.py ....................                               [ 94%]
test_syncspec.py .                                                       [ 94%]
test_traceimage.py .                                                     [ 95%]
test_utils.py ....                                                       [ 96%]
test_vcorr.py ..                                                         [ 97%]
test_wavemodel.py .                                                      [ 98%]
test_wavetilts.py ....                                                   [100%]

======================= 231 passed in 669.65s (0:11:09) ========================
```

Cooked uploaded.   Re-running dev suite.

**Comment by kbwestfall** [comment]
```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/70 TESTS (Masters ignored) ---
Failed tests:
    pypeit_tellfit: gemini_gnirs/32_SB_SXD
```

I was missing a file.  Running `./pypeit_test afterburn -i gemini_gnirs` to clean up now.

**Comment by kbwestfall** [comment]
Okay, all dev-suite tests now pass.  Here's the follow-up after re-running the gemini_gnirs afterburn:
```
--- PYPEIT DEVELOPMENT SUITE PASSED 6/6 TESTS (Masters ignored) ---
```

File: pypeit/core/wavecal/autoid.py [hunk 4]
```diff
@@ -973,144 +973,188 @@ def full_template(spec, par, ok_mask, det, binspectral, nsnippet=2, debug_xcorr=
     # Finish
     return wvcalib
```
Comment (jhennawi) [inline pypeit/core/wavecal/autoid.py:975]: Did you check that this documentation is in the parset? I spent a long time writing it and would prefer not to lose it.

File: pypeit/spectrographs/gemini_gnirs.py (hunk 3 above)
Comment (jhennawi) [reply pypeit/spectrographs/gemini_gnirs.py:336]: They probably should be converted into fraction of the detector. We do use this for X-shooter VIS, which can be binned. Worth checking what we are doing there.

File: pypeit/spectrographs/gemini_gnirs.py [hunk 5]
```diff
@@ -316,6 +316,11 @@ def order_spat_pos(self):
         else:
             msgs.error('Unrecognized disperser')
 
+#    @property
```
Comment (jhennawi) [inline pypeit/spectrographs/gemini_gnirs.py:319]: Where is this functionality now? How do I adjust this mtch_tol_spat parameter?

**Comment by kbwestfall** [comment]
After updating the pypeit file and the data in the dev-suite to the new data @jhennawi provided:

```
./pypeit_test gemini_gnirs -s 10_LB_SXD

...

--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS (Masters ignored) ---
```

I'll be finishing off the rest of Joe's comments this morning.

File: pypeit/core/wavecal/autoid.py (hunk 4 above)
Comment (kbwestfall) [reply pypeit/core/wavecal/autoid.py:975]: Done.

File: pypeit/spectrographs/gemini_gnirs.py (hunk 3 above)
Comment (kbwestfall) [reply pypeit/spectrographs/gemini_gnirs.py:336]: Punting to a different PR.

**Pull request merged**